File: backend/src/main/java/be/technobel/parsemaster/validation/validator/PositiveOrZeroValidator.java
package be.technobel.parsemaster.validation.validator;

import be.technobel.parsemaster.dto.ErrorFieldDTO;
import be.technobel.parsemaster.exception.InvalidParamException;
import be.technobel.parsemaster.validation.constraint.PositiveOrZero;
import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class PositiveOrZeroValidator implements ConstraintValidator<PositiveOrZero, Integer> {
  private String field;
  private String message;

  @Override
  public void initialize(PositiveOrZero annotation) {
    field = annotation.field();
    message = annotation.message();
  }

  @Override
  public boolean isValid(Integer value, ConstraintValidatorContext context) {
    if(value == null || value >= 0) return true;
    throw new InvalidParamException(new ErrorFieldDTO(field, message));
  }
}

File: src/main/kotlin/be/technobel/playzone/pl/validation/validators/PositiveOrZeroValidator.kt
package be.technobel.playzone.pl.validation.validators

import be.technobel.playzone.pl.rest.FormError
import be.technobel.playzone.pl.rest.InvalidParamException
import be.technobel.playzone.pl.validation.constraints.PositiveOrZero
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class PositiveOrZeroValidator : ConstraintValidator<PositiveOrZero, Int> {
	private lateinit var field: String
	private lateinit var message: String

	override fun initialize(constraintAnnotation: PositiveOrZero) {
		field = constraintAnnotation.field
		message = constraintAnnotation.message
	}

	override fun isValid(value: Int, context: ConstraintValidatorContext): Boolean {
		return if (value >= 0) true else throw InvalidParamException(
			FormError(field, "$field $message")
		)
	}
}

